### PR TITLE
Add context to before/after scenario hooks issue 5394

### PR DESCRIPTION
**Pull request by osmolyar**
## Proposed changes
Add context to before/after scenario hooks to resolve issue https://github.com/webdriverio/webdriverio/issues/5394

[//]: # Addressing issue https://github.com/webdriverio/webdriverio/issues/5394.

## Types of changes  Modeled on changes in https://github.com/webdriverio/webdriverio/pull/4545/files to add context parameter to hooks

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: tests/helpers/cucumber-hooks.conf.js [hunk 1]
```diff
@@ -27,9 +28,15 @@ exports.config = Object.assign({}, config, {
         }
         browser.Cucumber_Test = 1
     },
-    afterScenario: () => {
+    afterScenario: (world) => {
         browser.pause(30)
         browser.Cucumber_Test = -1
+        if (
+            browser.Cucumber_CurrentWorld !== world
+        ) {
```
Comment (christian-bromann) [inline tests/helpers/cucumber-hooks.conf.js:36]:
```suggestion
        if (browser.Cucumber_CurrentWorld !== world) {
```

File: tests/helpers/cucumber-hooks.conf.js [hunk 2]
```diff
@@ -7,9 +7,10 @@ exports.config = Object.assign({}, config, {
         await browser.pause(30)
         browser.Cucumber_Test = 0
     },
-    beforeScenario: () => {
+    beforeScenario: (world) => {
```
Comment (christian-bromann) [inline tests/helpers/cucumber-hooks.conf.js:10]: `world` should be the 5th parameter, no?

File: tests/helpers/cucumber-hooks.conf.js [hunk 3]
```diff
@@ -27,9 +28,15 @@ exports.config = Object.assign({}, config, {
         }
         browser.Cucumber_Test = 1
     },
-    afterScenario: () => {
+    afterScenario: (world) => {
```
Comment (christian-bromann) [inline tests/helpers/cucumber-hooks.conf.js:31]: same

File: packages/wdio-cucumber-framework/cucumber-framework.d.ts [hunk 4]
```diff
@@ -27,10 +27,10 @@ interface StepData {
 
 interface CucumberHookFunctions {
     beforeFeature?(uri: string, feature: CucumberHookObject, scenarios: CucumberHookObject[]): void;
-    beforeScenario?(uri: string, feature: CucumberHookObject, scenario: CucumberHookObject, sourceLocation: SourceLocation): void;
-    beforeStep?(step: StepData, context: any): void;
-    afterStep?(step: StepData, context: any, result: { error?: any, result?: any, passed: boolean, duration: number }): void;
-    afterScenario?(uri: string, feature: CucumberHookObject, scenario: CucumberHookObject, result: CucumberHookResult, sourceLocation: SourceLocation): void;
+    beforeScenario?(uri: string, feature: CucumberHookObject, scenario: CucumberHookObject, sourceLocation: SourceLocation, context?: World): void;
```
Comment (christian-bromann) [inline packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: Do we need to import this type (`World`)?

Comment (osmolyar) [reply packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: It is defined by Cucumber types and recognized by VSCode without importing anything.  Though still trying to test locally with no success due to "command not found: wdio".

Is this the right procedure?
1) npm run setup-full in webdriverio repo directory
2) run "npm pack" in webdriverio repo directory
3) run 'npm uninstall webdriverio" in project directory
3) run "npm install <path to webdriverio-monorepo.tgz file" in project directory

there is no wdio executable in node_modules\.bin after this.

File: tests/helpers/cucumber-hooks.conf.js (hunk 2 above)
Comment (osmolyar) [reply tests/helpers/cucumber-hooks.conf.js:10]: I would think so but was modeling it on before/afterStep which also only pass the newly added parameters:

`beforeStep: async function ({ step }, world) {
        await browser.pause(20)
        browser.Cucumber_Test += 2
        browser.Cucumber_CurrentStepText = step.step.text
        browser.Cucumber_CurrentWorld = world
    },
    afterStep: function ({ step }, world) {
        browser.pause(25)
        if (
            browser.Cucumber_CurrentStepText !== step.step.text ||
            browser.Cucumber_CurrentWorld !== world
        ) {
            throw new Error("step data doesn't match")
        }
        browser.Cucumber_Test = 1
    },`

File: packages/wdio-cucumber-framework/cucumber-framework.d.ts (hunk 4 above)
Comment (christian-bromann) [reply packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: please see our [contributing guidelines](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md#link-changes-to-your-current-project)

Comment (osmolyar) [reply packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: @christian-bromann , I tried the npm link procedure first but it keeps apparently linking to a global webdrivero, even after I uninstall -g webdriverio, not my modified C:\webderiverio where npm link was run:

 webdriverio@6.1.17
added 139 packages from 187 contributors in 10.23s
C:\osmolyar_qdtools\iris\latest\ManagementPortal\node_modules\webdriverio -> C:\Users\osmolyar\AppData\Roaming\npm\node_modules\webdriverio

Comment (osmolyar) [reply packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: Also tried just linking packages/wdio-cucumber-framework but when running 'npx wdio ...' it keeps saying 'command not found:  wdio'.  Even after doing 'npm i wdio'.

Comment (osmolyar) [reply packages/wdio-cucumber-framework/cucumber-framework.d.ts:30]: Managed to link @wdio/cucumber-framework but it doesn't seem to be working at all.  (In fact, can't even get context to console.log in before/afterStep where it is supposed to already be supported).  Will close PR for now.

File: tests/helpers/cucumber-hooks.conf.js (hunk 2 above)
Comment (osmolyar) [reply tests/helpers/cucumber-hooks.conf.js:10]: Managed to link @wdio/cucumber-framework but it doesn't seem to be working at all.  (In fact, can't even get context to console.log in before/afterStep where it is supposed to already be supported).  Will close PR for now.

**Comment by osmolyar** [comment]
Not working locally; closing for now

**Comment by christian-bromann** [comment]
> Not working locally; closing for now

I recommend to use our [smoke test](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md#run-e2e-experience-with-smoke-tests) to check if changes work. You can just try reproduce your desired scenario in our [cucumber smoke project](https://github.com/webdriverio/webdriverio/tree/master/tests/cucumber) and run it while you are changing files.

**Comment by osmolyar** [comment]
Thanks for the tips.  Am a little confused about how to use the cucumber smoke project because the thing that isn't 'working' is logging out the 'context' parameter in before/afterStep, before/afterScenario hood (before/afterStep even before the changes in this PR), and there is no config file with hooks in the smoke project.   In any case the cucumberTestrunner tests are also failing but it's not clear why:

PS C:\webdriverio> npm run test:smoke cucumberTestrunner

> webdriverio-monorepo@1.0.0 test:smoke C:\webdriverio
> babel-node ./tests/smoke.runner.js "cucumberTestrunner"


Running smoke tests...


Execution of 2 spec files started at 2020-06-12T16:42:03.397Z

[BABEL] Note: The code generator has deoptimised the styling of C:\webdriverio\packages\wdio-webdriver-mock-service\node_modules\lodash\lodash.js as it exceeds the max of 500KB.
[0-1] FAILED
[0-0] FAILED

Spec Files:      0 passed, 2 failed, 2 total (100% completed) in 00:00:13


========== LOG OUPUT wdio-0-0.log


========== LOG OUPUT wdio-0-1.log


========== LOG OUPUT wdio.log
2020-06-12T16:42:09.930Z INFO @wdio/local-runner: Start worker 0-0 with arg: cucumberTestrunner
2020-06-12T16:42:10.063Z INFO @wdio/local-runner: Start worker 0-1 with arg: cucumberTestrunner
2020-06-12T16:42:16.810Z DEBUG @wdio/local-runner: Runner 0-1 finished with exit code 1
2020-06-12T16:42:16.829Z DEBUG @wdio/local-runner: Runner 0-0 finished with exit code 1
2020-06-12T16:42:16.961Z INFO @wdio/local-runner: Shutting down spawned worker
2020-06-12T16:42:17.225Z INFO @wdio/local-runner: Waiting for 0 to shut down gracefully

Error: Smoke test failed
    at C:\webdriverio\tests\helpers\launch.js:39:11
    at cucumberTestrunner (C:\webdriverio\tests\smoke.runner.js:101:7)
    at runTests (C:\webdriverio\tests\smoke.runner.js:35:7)
    at runTests (C:\webdriverio\tests\smoke.runner.js:35:7)
    at C:\webdriverio\tests\smoke.runner.js:300:3
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! webdriverio-monorepo@1.0.0 test:smoke: `babel-node ./tests/smoke.runner.js "cucumberTestrunner"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the webdriverio-monorepo@1.0.0 test:smoke script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\osmolyar\AppData\Roaming\npm-cache\_logs\2020-06-12T16_42_19_295Z-debug.log

**Comment by christian-bromann** [comment]
> In any case the cucumberTestrunner tests are also failing but it's not clear why:

Is this with your change or without?

**Comment by osmolyar** [comment]
Both with and without changes.

**Comment by christian-bromann** [comment]
Can you try on a fresh checked out repository? These tests should pass (see CI pipeline).

**Comment by osmolyar** [comment]
Tried with a freshly cloned/checked out repository and the tests still fail.  (Also had to install @wdio/local-runner separately after npm install and npm run setup-full to run smoke tests.)

**Comment by osmolyar** [comment]
@christian-bromann just fyi the cucumber smoke tests pass when run in a standalone command prompt, but not from the same directory within VSCode.

**Comment by christian-bromann** [comment]
> when run in a standalone command prompt, but not from the same directory within VSCode.

Can you explain a bit further what you mean by that?

**Comment by osmolyar** [comment]
Sorry, the above comment is incorrect.  The command "npm run test:smoke cucumberTestrunner" results in the same Failed result both in a command prompt and in VSCode's terminal window.  It only says "All tests passed" when misspelled as "npm run test:smoke cucumberTestrunne" (presumably because then there are no tests to run).  In any case the PR should be closed as the code changes don't work.

**Comment by christian-bromann** [comment]
@osmolyar I think the change still makes a lot of sense? Let's re-open and we jump on a 1:1 session through the [WebdriverIO Office Hours](https://calendly.com/christian-bromann/webdriverio-office-hours?back=1&month=2020-06)?!

**Comment by osmolyar** [comment]
Ok- in a meeting for the next hour but otherwise anytime.  Thanks

**Comment by christian-bromann** [comment]
> but otherwise anytime

Please use the link to schedule a time. Thanks!

**Pull request merged**